### PR TITLE
Add Amber topology support to extract_dihedral_entropy.py

### PR DESCRIPTION
**Pull request by emmeburgin**
Adds support for Amber-format .prmtop (or .top) files when using extract_dihedral_entropy.py from the command line.

Previously this tool would fail when supplied an Amber topology file because mdtraj's format-agnostic load function does not support creating topology files without an associated trajectory. This change is a workaround that replaces the format-agnostic load function with the appropriate format-specific function whenever the topology filename ends in '.prmtop' or '.top'.

### CODE REVIEW

File: apps/extract_dihedral_entropy.py [hunk 1]
```diff
@@ -57,17 +57,25 @@ def get_resi_mapping(inds_file, top_File):
...
+    else:
+        structure = md.load(top_File)
+        n_resis = structure.top.n_residues
     n_dihedrals = inds.shape[0]
     resi_map = np.zeros(n_dihedrals)
     for i, n in enumerate(inds):
         atom_index = n[1]
-        resi_index = structure.top.atom(int(atom_index)).residue.index
+        if top_File[-7:] == '.prmtop' or top_File[-4:] == '.top':
+            resi_index = structure.atom(int(atom_index)).residue.index
```
Comment (sukritsingh) [inline apps/extract_dihedral_entropy.py]: While the file loading `if` statements make intuitive sense to me, could you elaborate here (maybe add a comment) why you're having to add this conditional at the `Topology` object level? By this point it's just MDtraj manipulation, right? Maybe just adding a comment here for posterity is sufficient. Thanks so much!

File: apps/extract_dihedral_entropy.py [hunk 2]
```diff
@@ -103,10 +111,16 @@ def save_dihedral_entropy(entropy_values, resi_map, output_name):
...
+        structure = md.load_prmtop(top_file)
+    else:
+        structure = md.load(top_file)
     resSeq_list = np.zeros(resi_list.shape[0])
     for i, n in enumerate(resi_list):
-        resSeq_list[i] = structure.top.residue(int(n)).resSeq
+        if top_file[-7:] == '.prmtop' or top_file[-4:] == '.top':
+            resSeq_list[i] = structure.residue(int(n)).resSeq
+        else:
+            resSeq_list[i] = structure.top.residue(int(n)).resSeq
```
Comment (sukritsingh) [inline apps/extract_dihedral_entropy.py]: See above comment about clarifying with comments

**Review by sukritsingh — COMMENTED** [review]
This is a great, clean change to add `prmtop` compatibility to the `extract_dihedral_entropy` scripts! I've add a comment about a couple of locations where a clarifying comment would be helpful for posterity, but that's it otherwise!

**Comment by emmeburgin** [comment]
Definitely, sorry for neglecting to comment this before. I have added comments explaining the changes to this pull request. This is addressed in the comments, but to answer this question:

> why you're having to add this conditional at the Topology object level? By this point it's just MDtraj manipulation, right?

MDTraj's `load` function is format-agnostic, but only for trajectory files. In this case, the modified functions aren't actually using trajectory (i.e., coordinate) information, only topology information. Adding support for Amber-style .prmtop files requires that we circumvent the `load` function and instead load only the topology information directly, but this produces a different MDTraj object: a topology instead of a trajectory with a topology attribute. Because MDTraj doesn't have a format-agnostic topology loading function, handling it on a case-by-case basis depending on file extension is the best available workaround.

**Comment by sukritsingh** [comment]
Ah I see now the issue you're talking about! Yes if MDTraj only returns Topology objects that way, but if you want a trajectory object you'd have to initialize a `Trajectory` with the already loaded `Topology` object.

> Definitely, sorry for neglecting to comment this before. I have added comments explaining the changes to this pull request.

No worries at all! These comments are helpful and informative. 

My only concern is around this bloating very quickly where we would have a bajillion `if` statements for each individual topology type if this comes up again beyond Amber specific formats. 

What do you think about a `load_topology_directly` function that uses `getattr`? Something like this: 
```
def load_topology_directly(filename):
    ext = os.path.splitext(filename)[1].lower()[1:]
    loader_fn = getattr(md, f'load_{ext}', None)

    topology = loader_fn(filename)

    return topology
```

and then using that function? You could still do the `if` `else` statement, or a `try` block that passes it to this function if `md.load` fails.

I'm happy to try a stab at this as well if this is a blocker for you, but I was just thinking about how to generalize this beyond Amber-specific topology files.

**Comment by emmeburgin** [comment]
I definitely share your instinct towards avoiding bloat of this sort, and building in some generality was something I considered myself. However, taking a look at the [format-specific loading functions](https://mdtraj.org/1.9.4/load_functions.html#format-specific-loading-functions) available in MDTraj, I believe that `prmtop` files are the only supported format that contains only a topology and not a trajectory. In other words, this problem is actually Amber-specific, and there are no other formats where this would come up (and that MDTraj would be able to support).

Admittedly your loader function is prettier, but in function I think it's exactly the same as a `prmtop`-specific `if` block. The advantage would be reducing bloat somewhat at the cost of readability, but with no benefit in function. Since this is your project, I leave that tradeoff to your discretion!

**Comment by sukritsingh** [comment]
> I believe that prmtop files are the only supported format that contains only a topology and not a trajectory. In other words, this problem is actually Amber-specific, and there are no other formats where this would come up (and that MDTraj would be able to support).

Oh yeah, really good point here. I didn't even put that together that this would only be a problem with the Amber type files.

I do like the idea of loader functions, and generalizing, but it sounds like this is the only relevant edge case, so I think this strategy is fine for now and I can revisit if somehow suddenly other formats become specific. 

The code passes the eye test for me - have you had a chance to test this code out in some of your own files and usecases? I don't have a `.prmtop` file handy to test with, but I can probably get one

**Comment by emmeburgin** [comment]
Yes, I have tested this out with a .prmtop file and it appears to have gone smoothly! I'd be happy to provide one for you to test with if you'd like to try it out yourself.

**Comment by sukritsingh** [comment]
Fantastic! If you'd be willing to upload it here in the thread or something? I can push it in along with some other "files for testing" in a future PR (particularly if others besides myself are using this!)

Thanks so much! This LGTM - happy to merge this in

**Comment by emmeburgin** [comment]
Sure, here's a `prmtop` file with a corresponding `rst7` coordinate file for _Aequorea victoria_ GFP. The coordinate file is just a single frame, so not useful for any actual analysis, but works as a test case for the code (`collect_cards.py` with these changes handles it fine).

[avgfp.zip](https://github.com/user-attachments/files/19032802/avgfp.zip)

**Comment by sukritsingh** [comment]
Amazing thank you!
